drivers/gpio/gpio_ad5592.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2023 Grinn
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_ad5592_gpio
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/drivers/mfd/ad5592.h>
+
+struct gpio_ad5592_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	const struct device *mfd_dev;
+};
+
+struct gpio_ad5592_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	uint8_t gpio_val;
+	uint8_t gpio_out;
+	uint8_t gpio_in;
+	uint8_t gpio_pull_down;
+};
+
+static int gpio_ad5592_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_ad5592_config *config = dev->config;
+	struct gpio_ad5592_data *drv_data = dev->data;
+	uint16_t data;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	ret = mfd_ad5592_read_reg(config->mfd_dev,
+				  AD5592_REG_GPIO_INPUT_EN, drv_data->gpio_in, &data);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*value = (uint32_t)data;
+
+	return 0;
+}
+
+static int gpio_ad5592_port_set_bits_raw(const struct device *dev,
+					  gpio_port_pins_t pins)
+{
+	struct gpio_ad5592_data *data = dev->data;
+	const struct gpio_ad5592_config *config = dev->config;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	data->gpio_val |= (uint8_t)pins;
+
+	return mfd_ad5592_write_reg(config->mfd_dev, AD5592_REG_GPIO_SET, data->gpio_val);
+}
+
+static int gpio_ad5592_port_clear_bits_raw(const struct device *dev,
+					    gpio_port_pins_t pins)
+{
+	struct gpio_ad5592_data *data = dev->data;
+	const struct gpio_ad5592_config *config = dev->config;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	data->gpio_val &= ~(uint8_t)pins;
+
+	return mfd_ad5592_write_reg(config->mfd_dev, AD5592_REG_GPIO_SET, data->gpio_val);
+}
+
+static inline int gpio_ad5592_configure(const struct device *dev,
+					 gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct gpio_ad5592_data *data = dev->data;
+	const struct gpio_ad5592_config *config = dev->config;
+	uint8_t val;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	if (pin >= AD5592_PIN_MAX) {
+		return -EINVAL;
+	}
+
+	val = BIT(pin);
+	if ((flags & GPIO_OUTPUT) != 0U) {
+		data->gpio_in &= ~val;
+		data->gpio_out |= val;
+
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			ret = gpio_ad5592_port_set_bits_raw(
+				dev, (gpio_port_pins_t)BIT(pin));
+			if (ret < 0) {
+				return ret;
+			}
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			ret = gpio_ad5592_port_clear_bits_raw(
+				dev, (gpio_port_pins_t)BIT(pin));
+			if (ret < 0) {
+				return ret;
+			}
+		}
+
+		ret = mfd_ad5592_write_reg(config->mfd_dev,
+					   AD5592_REG_GPIO_OUTPUT_EN, data->gpio_out);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = mfd_ad5592_write_reg(config->mfd_dev,
+					   AD5592_REG_GPIO_INPUT_EN, data->gpio_in);
+		if (ret < 0) {
+			return ret;
+		}
```

Comment:
This is repeated in the else, please move it outside the if/else

Reply:
Fixed.